File: services/app/lib/akashi/siws.ex
defmodule Akashi.SIWS do
  @moduledoc false
  alias Akashi.Accounts

  def verify_signature(%{address: address, message: message, signature: signature}) do
    request =
      Req.post!(
        "http://localhost:3000/siws",
        json: %{
          message: message,
          address: address,
          signature: signature
        }
      )

    case request.status do
      200 ->
        if Map.get(request.body, "verified") && Map.get(request.body, "verified") == true do
          IO.puts("Verified")
          {:ok, user} = Accounts.create_user_if_not_exists(address)
          user
        else
          IO.puts("Not verified")
          nil
        end

      _ ->
        IO.puts("Error")
        nil
    end
  end
end

File: app/ports/src/siws.ts
import { Request, Response } from "express";
import { SIWS } from "@web3auth/sign-in-with-solana";
import bs58 from "bs58";

async function SIWSHandler(req: Request, res: Response): Promise<void> {
  try {
    const message = JSON.parse(req.body.message);
    const header = message.header;
    const payload = message.payload;

    const signatureData = JSON.parse(req.body.signature);
    const signature_str = Uint8Array.from(Buffer.from(signatureData.signature.data));

    const signature = {
      t: "sip99",
      s: bs58.encode(signature_str)
    };

    const msg = new SIWS({ header, payload });
    const resp = await msg.verify({ payload, signature });

    if (resp.success == true) {
      console.log("Signature verified for address: ", payload.address);
      res.status(200).json({ verified: true });
    } else {
      console.log("Signature NOT verified for address: ", payload.address);
      res.status(200).json({ verified: false });
    }
  } catch (e: any) {
    console.error("Error occurred:", e.message);
    res.status(400).json({ error: e.message });
  }
}

export default SIWSHandler;
